### ISSUE TITLE
Implement support for Kubelet Dynamic Configuration

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubeadm#28]
_From @luxas on October 5, 2016 19:26_

We should use Kubelet Dynamic Settings by pushing values given to `kubeadm init` to the apiserver when started, and that will make all kubelets in the cluster reload and pick up those configuration values.

That's much better than having to modify every `10-kubeadm.conf` systemd dropin file in the whole cluster.

_Copied from original issue: kubernetes/kubernetes#34132_

**Comment by mikedanese** [comment]
_From @luxas on October 5, 2016 19:28_

cc @kubernetes/sig-cluster-lifecycle

**Comment by mikedanese** [comment]
_From @errordeveloper on October 5, 2016 20:25_

Is this feature already usable?

**Comment by mikedanese** [comment]
_From @luxas on October 5, 2016 20:29_

I don't know actually, but I think an alpha version of it exists

cc @mtaufen @vishh @mikedanese @dchen1107 @yujuhong

**Comment by mikedanese** [comment]
_From @axsuul on October 7, 2016 6:58_

Is this related to customizing arguments for the kubernetes API server? I'm trying to customize the secure port it listens on (`--secure-port=8081`) but I'm not sure how to do it when setting up the cluster with `kubeadm`.

**Comment by mikedanese** [comment]
_From @errordeveloper on October 7, 2016 7:15_

James, no this is not related. To your question, please try editing
/etc/kubernetes/manifests/kube-apiserver.json, but please create another
issue if you have more questions.

**Comment by mikedanese** [comment]
_From @axsuul on October 7, 2016 7:58_

@errordeveloper Thanks, I've opened up issue #34307 with more questions

**Comment by mikedanese** [comment]
_From @mtaufen on October 7, 2016 17:3_

There is an alpha version of the dynamic settings feature, but it still needs a lot of work. I'm going to be mostly focused on GCI-related things for v1.5, but am planning be giving it high-priority focus for v1.6. I wouldn't recommend building on top of dynamic settings just yet, as it will undergo some significant changes (e.g. we will move to telling a node which configmap to use via annotations rather than creating configmaps with a per-node naming convention). 

As it stands you would have to create a configmap for every node in the cluster anyway, and based on @luxas's first comment in this issue, it looks like you are trying to avoid per-node work, so that might be another reason to wait. Granted, this per-node work is still probably less effort than modifying systemd conf across a cluster...

**Comment by pires** [comment]
#64 and #42 are potential use-cases for this.

**Comment by luxas** [comment]
We are blocked on this one from other persons/sigs: https://github.com/kubernetes/kubernetes/issues/27980

**Comment by jamiehannaford** [comment]
This might be possible for v1.8. 
Feature issue: https://github.com/kubernetes/features/issues/281
alpha PR in review: https://github.com/kubernetes/kubernetes/pull/46254

**Comment by luxas** [comment]
@mtaufen Will the alpha version of Kubelet Dynamic Configuration be easily upgradeable?
I mean, will it be of beta quality but alpha just because of the ladder and policy to not just enable things by default?

kubeadm could be a good way to test these things out on real clusters during the cycle and then we can decide whether we want to enable it or not in v1.8

**Comment by mtaufen** [comment]
It depends on how confident we are that the API types for the Kubelet won't change after 1.8. It's too early to say right now.

**Comment by luxas** [comment]
@mtaufen Kubelet Dynamic Configuration and the kubelet API type will be alpha in v1.8, but beta is targeted for v1.9, right?
Bumping kubeadm here to use it in v1.9

**Comment by mtaufen** [comment]
Yes, or even GA in 1.9. It's probably most reasonable to focus on a "clean" API in kubeletconfig v2 and just canonize the existing one that people depend on in v1 ASAP, so that we can start thinking about v2.

**Comment by luxas** [comment]
So we're gonna start implementing this _now_.
@xiangpengzhao has volunteered to take this task on, woot and thanks!

For reference, here are some docs we've worked on earlier:
 - https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/
 - https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

We'll add a new feature gate to kubeadm: `DynamicKubeletConfig` which will enable the `DynamicKubeletConfig` feature gate on the API server as well.

We'll add a new struct to our MasterConfiguration API:
```go
type KubeletConfiguration struct {
      BaseConfig *kubeletconfig.KubeletConfiguration
}
```

The contents of `BaseConfig` if empty will be populated with defaults for the current k8s version (the options we now have populated as flags [here](https://github.com/kubernetes/release/blob/master/debian/xenial/kubeadm/channel/stable/etc/systemd/system/kubelet.service.d/post-1.8/10-kubeadm.conf) and then be uploaded to a ConfigMap named something like `kubelet-base-config-1.X` where `X` is the minor k8s version.

The idea here is that this ConfigMap will be shared as the base config for all kubelets (probably of a certain minor version, if needed) in the cluster

After the ConfigMap is created,  the UID can be set on the Node's `.spec.configSource.configMapRef.uid` etc.

We still need to wait for Node Authorizer support for Dynamic Kubelet Config and the Kubelet Config API to graduate to beta (https://github.com/kubernetes/kubernetes/pull/53833,  dependent on other PRs). But these efforts will happen in parallel to some degree.

We need to see how we can deal with upgrading kubelets smoothly as well etc. but let's start with something at least so we can test this out for real :)

**Comment by xiangpengzhao** [comment]
@luxas thanks for the details! Will dig into this and start working ASAP :)

**Comment by luxas** [comment]
We have some things to fix up yet:
1. Fix a panic https://github.com/kubernetes/kubeadm/issues/553 (https://github.com/kubernetes/kubernetes/pull/56008)
2. ClusterDNS should be set conditionally based on `cfg.Networking.ServiceSubnet` (https://github.com/kubernetes/kubernetes/pull/56013)
3. Re-engineer the `kubeadm join` logic to wait for the kubelet to create `/etc/kubernetes/kubelet.conf` and use those credentials to PATCH the node
4. Write the the marshalled kubeletconfig object to `/var/lib/kubelet/config/init/kubelet` so that the kubelet will start up with the right params on init/join. The only params expected in the kubelet command-line after this is `kubelet --init-config-dir /var/lib/kubelet/config/init --dynamic-config-dir /var/lib/kubelet/config/dynamic`

**Comment by mtaufen** [comment]
FYI `--init-config-dir` is likely to change a bit, wrt this thread: https://github.com/kubernetes/kubernetes/pull/55718#discussion_r151893476

**Comment by luxas** [comment]
Yeah, we'll accommodate accordingly when we move out of alpha with this on the kubeadm side.
Thanks for keeping us updated @mtaufen

**Comment by luxas** [comment]
FYI @timothysc ^
We'll do a revised version of https://github.com/kubernetes/kubeadm/blob/master/docs/design/design_v1.8.md soon-ish (with @fabriziopandini?) and include this in there.
Everything here is alpha gated in v1.9, planned for beta in v1.10

**Comment by luxas** [comment]
I think all the tasks needed to be performed were merged last week, so closing this as fixed for an alpha state now. The design of this feature is covered in design doc update of v1.9 #573.

We'll open new issues later for graduating this feature to beta on the kubeadm side.

Thanks @xiangpengzhao for the awesome work on this :tada:!

**Comment by bronger** [comment]
Is https://github.com/kubernetes/kubeadm/issues/83 supposed to work for v1.11? Because I still need to set
```
KUBELET_EXTRA_ARGS=--cluster-dns=10.67.97.10
```
in `/etc/sysconfig/kubelet` so that the IP of the DNS server in the pods is correct. I called:
```
kubeadm init --apiserver-cert-extra-sans kubmaster,kubmaster.mygreatdomain.de \
    --service-cidr 10.67.97.0/24
```
Without setting `KUBELET_EXTRA_ARGS`, the DNS is looked for at 10.96.0.10.

**Comment by neolit123** [comment]
@bronger 
i cannot comment on the topic if the kubeadm flag (`service-cidr`) **should** somehow translate to the DNS IP in the kubelet.

but assuming you have transitioned to the kubeadm config `v1alpha2` you can use the following:
```
nodeRegistration:
  kubeletExtraArgs:
    cluster-dns: 10.67.97.10
```
this will help you to not have to update the `kubelet` file.

ref: https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/

**Comment by luxas** [comment]
@bronger check the `/var/lib/kubelet/config.yaml` file, cluster-dns should be set correctly there in v1.11

**Comment by luxas** [comment]
@bronger please open a new issue with all the required information to debug that. Might be a bug.

**Comment by zerkms** [comment]
@bronger have you reported one? could you refer to it please, I can confirm on a latest stable kubeadm `kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:14:39Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}` the problem still is reproducible.

**Comment by bronger** [comment]
No, I haven't so far.  Long vacation and business trip came into way.  @luxas Does it may sense to just re-open https://github.com/kubernetes/kubeadm/issues/83?

**Comment by zerkms** [comment]
I think #83 perfectly fits, the new one would be identical
